### PR TITLE
bt-422: Create chat list screen

### LINKED ISSUE #422: Create chat list screen
**User story:**
As a user, I want to have a chat list, so that I am able to find the dialogue with potential employers.

**Acceptance criteria:**
- There should be the “Chats” name of the page in the upper left corner
- The chat list must include an input field for the search field on the top
- There should be a placeholder for the search field 
- A list of users’ chats should be displayed below the search field
- When the user writes a search query, only chats for my search query are displayed
- When the user selects chat (tap on it) it should somehow be highlighted and open in a new window
- Every chat should be displayed with the user’s profile photo, user name, the last message and the last message time of delivery
- When the user  taps on the chat it should open in a new screen

Image

### PR DESCRIPTION
**Pull request by anogire**
The buttons are for testing purposes only.
I have sorted by the latest messages.
Also, the date is updated every minute (in the video at the end).

But I'm not sure about the structure of the store itself, and I'm definitely not sure about the socket.
And I don't understand how I can test it if I need to do it now, in this task.

https://github.com/BinaryStudioAcademy/bsa-2023-bsa-talents/assets/70234990/95fd635e-69f9-4c3d-929a-0103958a78eb

### CODE REVIEW

File: mobile/src/bundles/employer/screens/chat/chat.tsx [hunk 1]
```diff
@@ -1,9 +1,153 @@
...
+    const sortedChats = useMemo(() => {
+        if (!filteredChats) {
+            return [];
+        }
+
+        return [...filteredChats].sort((a, b) => {
+            const dateA = new Date(a.lastMessageDate ?? '').getTime();
+            const dateB = new Date(b.lastMessageDate ?? '').getTime();
+            return dateB - dateA;
+        });
```
Comment (MaksymGavatuk) [inline mobile/src/bundles/employer/screens/chat/chat.tsx]: As for me, better to move it to helper

File: mobile/src/bundles/employer/constants/constants.ts [hunk 2]
```diff
@@ -0,0 +1,126 @@
+//TODO delete mock-data
+import { type ChatData } from '~/bundles/employer/types/types';
+
+const imageAvatar1 =
```
Comment (nikita-remeslov) [inline mobile/src/bundles/employer/constants/constants.ts]: not critical for mock, but we dont use such naming

`firstImageAvatar` - looks better

**Review by nikita-remeslov — APPROVED** [review]

File: mobile/src/bundles/employer/components/chat-list-item/chat-list-item.tsx [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+const ChatListItem: React.FC<Properties> = ({ item, isSelected, onSelect }) => {
+    const { userId, username, avatar, lastMessage, lastMessageDate } = item;
+    const lastMessageTimeDelivery = useRealTimeElapsed(lastMessageDate ?? '');
+
+    const itemAvatar = avatar ? (
+        <Image source={{ uri: avatar }} style={styles.defaultAvatar} />
+    ) : (
+        <View style={styles.defaultAvatar} />
+    );
+
```
Comment (katerynakostikova) [inline mobile/src/bundles/employer/components/chat-list-item/chat-list-item.tsx]: let's use avatar component

File: mobile/src/bundles/employer/components/chat-list-item/chat-list-item.tsx [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        onSelect(userId);
+    }, [onSelect, userId]);
+
+    return (
+        <Pressable onPress={handleListItemSelect}>
+            <View
+                style={[
+                    globalStyles.flexDirectionRow,
+                    globalStyles.p10,
+                    isSelected && styles.selectedItem,
```
Comment (katerynakostikova) [inline mobile/src/bundles/employer/components/chat-list-item/chat-list-item.tsx]: if chat is selected, then we'll just open a new screen, I guess we don't need that prop and styles

File: mobile/src/bundles/employer/constants/constants.ts [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    'https://img.freepik.com/free-psd/3d-illustration-person-with-sunglasses_23-2149436188.jpg?w=1380&t=st=1695064264~exp=1695064864~hmac=eb1d12fc35d9099271c9530148d45aa9301b83fa2b54c615a3739f8025a4cd95';
+const imageAvatar2 =
+    'https://img.freepik.com/free-psd/3d-illustration-person-with-sunglasses_23-2149436178.jpg?w=1380&t=st=1695169984~exp=1695170584~hmac=9091c0e95eb97e5335a9384e2d77016ed9ed1d3c6777c082bb8dc1a5c5e7abcd';
+
+const listItems: ChatData[] = [
+    {
+        talentId: 'u1',
+        talentName: 'Talent Name 1',
+        talentAvatar: imageAvatar1,
+        data: [
```
Comment (katerynakostikova) [inline mobile/src/bundles/employer/constants/constants.ts]: how chat will look like for talent role?

File: mobile/src/bundles/employer/screens/chat/chat.tsx [hunk 6]
```diff
@@ -1,9 +1,153 @@
...
+import { globalStyles } from '~/bundles/common/styles/styles';
+import { ChatListItem, Search } from '~/bundles/employer/components/components';
+import { actions as chatActions } from '~/bundles/employer/store';
+import { type ChatListItemType } from '~/bundles/employer/types/types';
+
+import { styles } from './styles';
 
 const Chat: React.FC = () => {
-    return <Text>Employer Chat screen</Text>;
+    const { currentUserData } = useAppSelector(({ auth }) => auth);
```
Comment (katerynakostikova) [inline mobile/src/bundles/employer/screens/chat/chat.tsx]: Let's name the screen ChatList, Chat will be for single dialog

File: mobile/src/bundles/employer/screens/chat/chat.tsx [hunk 7]
```diff
@@ -1,9 +1,153 @@
...
+            talentId: 'u2',
+            talentName: 'Talent Name 2',
+            message: 'New message text2',
+        };
+
+        void dispatch(chatActions.sendMessage(newMessage));
+    }, [dispatch, currentUserData?.id]);
+
+    const [searchQuery, setSearchQuery] = useState('');
+    const [selectedItemId, setSelectedItemId] = useState<string | null>(null);
```
Comment (katerynakostikova) [inline mobile/src/bundles/employer/screens/chat/chat.tsx]: i think we don't need to store selectedItemId

File: mobile/src/bundles/employer/screens/chat/chat.tsx [hunk 8]
```diff
@@ -1,9 +1,153 @@
...
+                item={item}
+                isSelected={item.userId === selectedItemId}
+                onSelect={handleChatSelect}
+            />
+        );
+    };
+
+    const handleChatSelect = useCallback(
+        (userId: string): void => {
+            setSelectedItemId(userId);
```
Comment (katerynakostikova) [inline mobile/src/bundles/employer/screens/chat/chat.tsx]: TODO navigate to separate screen

File: mobile/src/bundles/employer/store/slice.ts [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    dataStatus: DataStatus.IDLE,
+    chatData: listItems,
+};
+
+const { reducer, actions, name } = createSlice({
+    initialState,
+    name: 'employers',
+    reducers: {},
+    extraReducers(builder) {
+        builder.addCase(clearEmployerStore, (state) => {
```
Comment (katerynakostikova) [inline mobile/src/bundles/employer/store/slice.ts]: again question about chat for talent

File: mobile/src/framework/store/store.package.ts [hunk 10]
```diff
@@ -9,14 +9,16 @@ import flipper from 'redux-flipper';
...
 import { reducer as talentsReducer } from '~/bundles/talent/store';
 import { talentApi } from '~/bundles/talent/talent';
 import { type Config } from '~/framework/config/config';
 import { notifications } from '~/framework/notifications/notifications';
-import { storage } from '~/framework/storage/storage';
+import { socketMiddleware, storage } from '~/framework/storage/storage';
 
 type RootReducer = {
     auth: ReturnType<typeof authReducer>;
+    employers: ReturnType<typeof employersReducer>;
```
Comment (katerynakostikova) [inline mobile/src/framework/store/store.package.ts]: can we make chat reducer, so that it is both for talent and employer?

**Comment by anogire** [comment]
https://github.com/BinaryStudioAcademy/bsa-2023-bsa-talents/assets/70234990/5e2388bd-12b1-4fc4-9059-74e5e3f62d0d


Screenshot 2023-09-20 at 23 02 27

File: mobile/src/bundles/chat/constants/constants.ts [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+
+const SECOND_TALENT_ID = 'talent2';
+const SECOND_TALENT_NAME = 'Second Talent';
+const SECOND_TALENT_AVATAR =
+    'https://img.freepik.com/free-psd/3d-illustration-person-with-sunglasses_23-2149436178.jpg?w=1380&t=st=1695169984~exp=1695170584~hmac=9091c0e95eb97e5335a9384e2d77016ed9ed1d3c6777c082bb8dc1a5c5e7abcd';
+
+const EMPLOYER_ID = 'employerId';
+const EMPLOYER = 'Facebook';
+
+const chatListItems: Record<string, ChatDataRequestDto[]> | null = {
```
Comment (katerynakostikova) [inline mobile/src/bundles/chat/constants/constants.ts]: with CAPS

File: mobile/src/bundles/common/types/navigation/root-navigation-parameter-list.type.ts [hunk 12]
```diff
@@ -1,9 +1,12 @@
 import { type RootScreenName } from '~/bundles/common/enums/enums';
 
+import { type ChatNavigationProperties } from './chat-navigation-properties.type';
+
 type RootNavigationParameterList = {
     [RootScreenName.AUTH_ROOT_ROUTE]: undefined;
     [RootScreenName.MAIN_ROOT_ROUTE]: undefined;
     [RootScreenName.ONBOARDING_ROOT_ROUTE]: undefined;
+    [RootScreenName.CHAT_ROOT_ROUTE]: ChatNavigationProperties;
```
Comment (katerynakostikova) [inline mobile/src/bundles/common/types/navigation/root-navigation-parameter-list.type.ts]:
```suggestion
    [RootScreenName.CHAT]: ChatNavigationProperties;
```

**Review by katerynakostikova — APPROVED** [review]

**Pull request merged**